File: src/cleansia_android/partner-app/src/main/java/cz/cleansia/partner/features/orders/WorkContractSheet.kt
package cz.cleansia.partner.features.orders

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Info
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.pluralStringResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import cz.cleansia.core.format.formatOrderDateRange
import cz.cleansia.core.format.formatOrderDateTime
import cz.cleansia.core.format.formatOrderPrice
import cz.cleansia.core.ui.components.CleansiaPrimaryButton
import cz.cleansia.core.ui.components.HtmlContentView
import cz.cleansia.core.ui.state.ActionState
import cz.cleansia.core.ui.theme.Spacing
import cz.cleansia.partner.R
import cz.cleansia.partner.data.orders.WorkContract
import cz.cleansia.partner.data.orders.WorkContractAcceptanceFacts
import cz.cleansia.partner.data.orders.WorkContractJobFacts
import cz.cleansia.partner.ui.theme.CleansiaPartnerTheme
import java.util.Locale

/**
 * The contract for work, on one screen with the gesture that accepts it: the job's facts, the text,
 * and the swipe beneath. A take and a standalone acceptance differ only in what the swipe calls; a
 * read shows the stored facts and the accepted version with no swipe at all.
 *
 * Every open re-loads — the facts are the server's word at that moment, never the list row's.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun WorkContractSheet(
    request: WorkContractRequest,
    onDismiss: () -> Unit,
    onOutcome: (WorkContractOutcome) -> Unit,
    viewModel: WorkContractSheetViewModel = hiltViewModel(),
) {
    val uiState by viewModel.uiState.collectAsStateWithLifecycle()
    val actionState by viewModel.actionState.collectAsStateWithLifecycle()
    val notice by viewModel.notice.collectAsStateWithLifecycle()
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    val submitting = actionState is ActionState.Submitting

    LaunchedEffect(viewModel, request) { viewModel.open(request) }
    LaunchedEffect(viewModel) { viewModel.outcome.collect(onOutcome) }

    ModalBottomSheet(
        onDismissRequest = { if (!submitting) onDismiss() },
        sheetState = sheetState,
        containerColor = MaterialTheme.colorScheme.surface,
    ) {
        WorkContractSheetContent(
            request = request,
            uiState = uiState,
            notice = notice,
            submitting = submitting,
            onAccept = viewModel::accept,
            onRetry = viewModel::retry,
            onClose = onDismiss,
        )
    }
}

@Composable
fun WorkContractSheetContent(
    request: WorkContractRequest,
    uiState: WorkContractUiState,
    notice: WorkContractNotice?,
    submitting: Boolean,
    onAccept: () -> Unit,
    onRetry: () -> Unit,
    onClose: () -> Unit,
) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .fillMaxHeight()
            .padding(horizontal = Spacing.M)
            .navigationBarsPadding(),
    ) {
        Text(
            text = (uiState as? WorkContractUiState.Loaded)?.contract?.title
                ?: stringResource(R.string.work_contract_title),
            style = MaterialTheme.typography.titleLarge.copy(fontWeight = FontWeight.Bold),
            color = MaterialTheme.colorScheme.onSurface,
        )
        Spacer(Modifier.height(Spacing.S))

        when (uiState) {
            WorkContractUiState.Loading -> Box(
                modifier = Modifier.fillMaxSize(),
                contentAlignment = Alignment.Center,
            ) { CircularProgressIndicator() }

            WorkContractUiState.Error -> SheetMessage(
                text = stringResource(R.string.work_contract_load_error),
                ctaLabel = stringResource(R.string.retry),
                onCta = onRetry,
            )

            WorkContractUiState.Unavailable -> SheetMessage(
                text = stringResource(R.string.error_legal_document_not_found),
                ctaLabel = stringResource(R.string.close),
                onCta = onClose,
            )

            is WorkContractUiState.Loaded -> LoadedContract(
                request = request,
                contract = uiState.contract,
                notice = notice,
                submitting = submitting,
                onAccept = onAccept,
                onClose = onClose,
            )
        }
    }
}

@Composable
private fun LoadedContract(
    request: WorkContractRequest,
    contract: WorkContract,
    notice: WorkContractNotice?,
    submitting: Boolean,
    onAccept: () -> Unit,
    onClose: () -> Unit,
) {
    Column(modifier = Modifier.fillMaxSize()) {
        Text(
            text = stringResource(R.string.work_contract_version, contract.version),
            style = MaterialTheme.typography.labelMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        Spacer(Modifier.height(Spacing.S))
        JobFacts(facts = contract.facts)
        contract.acceptance?.let { acceptance ->
            Spacer(Modifier.height(Spacing.XS))
            AcceptanceFacts(acceptance = acceptance, renderedLanguage = contract.language)
        }
        if (notice == WorkContractNotice.TextUpdated) {
            Spacer(Modifier.height(Spacing.S))
            NoticeRow(text = stringResource(R.string.error_contract_text_mismatch))
        }
        Spacer(Modifier.height(Spacing.S))
        HorizontalDivider(color = MaterialTheme.colorScheme.outlineVariant)
        HtmlContentView(
            html = contract.contentHtml,
            modifier = Modifier
                .fillMaxWidth()
                .weight(1f),
        )
        HorizontalDivider(color = MaterialTheme.colorScheme.outlineVariant)
        Spacer(Modifier.height(Spacing.S))
        when (request) {
            is WorkContractRequest.Take, is WorkContractRequest.Accept -> SlideToCommit(
                idleLabel = stringResource(R.string.work_contract_swipe_to_accept),
                busyLabel = stringResource(R.string.work_contract_accepting),
                onCommit = onAccept,
                isBusy = submitting,
            )
            is WorkContractRequest.Read -> CleansiaPrimaryButton(
                text = stringResource(R.string.close),
                onClick = onClose,
            )
        }
        Spacer(Modifier.height(Spacing.M))
    }
}

@Composable
private fun JobFacts(facts: WorkContractJobFacts) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .background(MaterialTheme.colorScheme.surfaceVariant.copy(alpha = 0.5f), RoundedCornerShape(12.dp))
            .padding(Spacing.S),
        verticalArrangement = Arrangement.spacedBy(Spacing.XXS),
    ) {
        Row(modifier = Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
            Text(
                text = facts.orderNumber?.takeIf { it.isNotBlank() }
                    ?.let { stringResource(R.string.work_contract_order_number, it) }
                    ?: formatOrderDateRange(facts.cleaningDateTimeUtc, facts.estimatedMinutes),
                style = MaterialTheme.typography.titleSmall.copy(fontWeight = FontWeight.SemiBold),
                color = MaterialTheme.colorScheme.onSurface,
                modifier = Modifier.weight(1f),
            )
            Text(
                text = formatOrderPrice(facts.totalPrice, facts.currencyCode),
                style = MaterialTheme.typography.titleSmall.copy(fontWeight = FontWeight.Bold),
                color = MaterialTheme.colorScheme.primary,
            )
        }
        if (!facts.orderNumber.isNullOrBlank()) {
            FactLine(formatOrderDateRange(facts.cleaningDateTimeUtc, facts.estimatedMinutes))
        }
        facts.locationApproximate?.takeIf { it.isNotBlank() }?.let { FactLine(it) }
        FactLine(
            listOfNotNull(
                facts.rooms.takeIf { it > 0 }?.let { pluralStringResource(R.plurals.scope_rooms, it, it) },
                facts.bathrooms.takeIf { it > 0 }?.let { pluralStringResource(R.plurals.scope_baths, it, it) },
            ).joinToString(" · "),
        )
        val scope = facts.packages + facts.services + facts.extraSlugs.map { nameForExtraSlug(it) }
        if (scope.isNotEmpty()) FactLine(scope.joinToString(", "))
    }
}

@Composable
private fun FactLine(text: String) {
    if (text.isBlank()) return
    Text(
        text = text,
        style = MaterialTheme.typography.bodySmall,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
    )
}

@Composable
private fun AcceptanceFacts(acceptance: WorkContractAcceptanceFacts, renderedLanguage: String?) {
    Column(verticalArrangement = Arrangement.spacedBy(Spacing.XXS)) {
        Text(
            text = stringResource(
                R.string.work_contract_accepted_on,
                formatOrderDateTime(acceptance.acceptedOn),
                acceptance.documentVersion,
            ),
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        val accepted = acceptance.acceptedLanguage?.takeIf { it.isNotBlank() }
        if (accepted != null && renderedLanguage != null && !accepted.equals(renderedLanguage, ignoreCase = true)) {
            Text(
                text = stringResource(
                    R.string.work_contract_accepted_in_language,
                    Locale.forLanguageTag(accepted).getDisplayLanguage(Locale.getDefault()),
                ),
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
    }
}

@Composable
private fun NoticeRow(text: String) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .background(MaterialTheme.colorScheme.tertiaryContainer, RoundedCornerShape(12.dp))
            .padding(Spacing.S),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(Spacing.XS),
    ) {
        Icon(
            imageVector = Icons.Outlined.Info,
            contentDescription = null,
            tint = MaterialTheme.colorScheme.onTertiaryContainer,
            modifier = Modifier.size(18.dp),
        )
        Text(
            text = text,
            style = MaterialTheme.typography.bodySmall.copy(fontWeight = FontWeight.SemiBold),
            color = MaterialTheme.colorScheme.onTertiaryContainer,
        )
    }
}

@Composable
private fun SheetMessage(text: String, ctaLabel: String, onCta: () -> Unit) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(vertical = Spacing.XL),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(Spacing.M),
    ) {
        Text(
            text = text,
            style = MaterialTheme.typography.bodyLarge,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        CleansiaPrimaryButton(text = ctaLabel, onClick = onCta)
        Spacer(Modifier.height(Spacing.M))
    }
}

@Preview
@Composable
private fun WorkContractSheetPreview() {
    CleansiaPartnerTheme {
        WorkContractSheetContent(
            request = WorkContractRequest.Take("order-1"),
            uiState = WorkContractUiState.Loaded(
                WorkContract(
                    legalDocumentTextId = "text-1",
                    version = "2026-09-20",
                    language = "en",
                    title = "Contract for Work",
                    contentHtml = "<p>This contract for work is concluded between the customer and the cleaner.</p>",
                    facts = WorkContractJobFacts(
                        orderNumber = "CL-2026-0042",
                        cleaningDateTimeUtc = "2026-08-12T09:00:00Z",
                        estimatedMinutes = 180,
                        totalPrice = 1850.0,
                        currencyCode = "CZK",
                        locationApproximate = "Praha 4 · 14000",
                        rooms = 3,
                        bathrooms = 1,
                        services = listOf("Standard cleaning"),
                        packages = emptyList(),
                        extraSlugs = listOf("inside-oven"),
                    ),
                    acceptance = null,
                ),
            ),
            notice = WorkContractNotice.TextUpdated,
            submitting = false,
            onAccept = {},
            onRetry = {},
            onClose = {},
        )
    }
}

File: src/cleansia_ios/CleansiaPartner/Sources/Features/Orders/WorkContractSheet.swift
import CleansiaCore
import SwiftUI

/// The contract for work, on one screen with the gesture that accepts it: the job's facts, the text,
/// and the swipe beneath. A take and a standalone acceptance differ only in what the swipe calls; a
/// read shows the stored facts and the accepted version with no swipe at all.
///
/// Built per presentation, so every open re-loads — the facts are the server's word at that moment,
/// never the list row's.
struct WorkContractSheet: View {
    @StateObject private var vm: WorkContractSheetViewModel
    private let onDismiss: () -> Void
    private let onOutcome: (WorkContractOutcome) -> Void

    init(
        request: WorkContractRequest,
        client: PartnerOrderClient,
        onDismiss: @escaping () -> Void,
        onOutcome: @escaping (WorkContractOutcome) -> Void
    ) {
        _vm = StateObject(wrappedValue: WorkContractSheetViewModel(request: request, client: client))
        self.onDismiss = onDismiss
        self.onOutcome = onOutcome
    }

    private var submitting: Bool {
        vm.actionState.isSubmitting
    }

    var body: some View {
        NavigationStack {
            WorkContractSheetContent(
                request: vm.request,
                state: vm.state,
                notice: vm.notice,
                submitting: submitting,
                onAccept: { Task { await vm.accept() } },
                onRetry: { Task { await vm.load() } },
                onClose: onDismiss
            )
            .navigationTitle(vm.state.loadedContract?.title ?? L10n.WorkContract.title)
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button(L10n.close, action: onDismiss).disabled(submitting)
                }
            }
            .background(CleansiaColors.surface.ignoresSafeArea())
        }
        .interactiveDismissDisabled(submitting)
        .task { await vm.load() }
        .onReceive(vm.outcome) { onOutcome($0) }
    }
}

struct WorkContractSheetContent: View {
    let request: WorkContractRequest
    let state: WorkContractSheetState
    var notice: WorkContractNotice?
    var submitting = false
    var onAccept: () -> Void = {}
    var onRetry: () -> Void = {}
    var onClose: () -> Void = {}

    var body: some View {
        switch state {
        case .loading:
            ProgressView()
                .frame(maxWidth: .infinity, maxHeight: .infinity)
        case .error:
            SheetMessage(text: L10n.WorkContract.loadError, ctaLabel: L10n.retry, onCta: onRetry)
        case let .unavailable(error):
            SheetMessage(text: ApiErrorLocalizer().message(for: error), ctaLabel: L10n.close, onCta: onClose)
        case let .loaded(contract):
            LoadedContract(
                request: request,
                contract: contract,
                notice: notice,
                submitting: submitting,
                onAccept: onAccept,
                onClose: onClose
            )
        }
    }
}

private struct LoadedContract: View {
    @Environment(\.locale) private var locale
    let request: WorkContractRequest
    let contract: WorkContract
    let notice: WorkContractNotice?
    let submitting: Bool
    let onAccept: () -> Void
    let onClose: () -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: Spacing.s) {
            Text(L10n.WorkContract.version(contract.version))
                .font(CleansiaTypography.labelMedium)
                .foregroundColor(CleansiaColors.onSurfaceVariant)
            JobFactsCard(facts: contract.facts, locale: locale)
            if let acceptance = contract.acceptance {
                AcceptanceFacts(acceptance: acceptance, renderedLanguage: contract.language, locale: locale)
            }
            if let notice {
                NoticeRow(text: notice.message)
            }
            Divider().overlay(CleansiaColors.outlineVariant)
            HtmlContentView(html: contract.contentHtml)
                .frame(maxWidth: .infinity, maxHeight: .infinity)
            Divider().overlay(CleansiaColors.outlineVariant)
            gesture
        }
        .padding(.horizontal, Spacing.m)
        .padding(.top, Spacing.s)
        .padding(.bottom, Spacing.m)
    }

    @ViewBuilder
    private var gesture: some View {
        switch request {
        case .take, .accept:
            SlideToConfirm(
                idleLabel: L10n.WorkContract.swipeToAccept,
                busyLabel: L10n.WorkContract.accepting,
                isBusy: submitting,
                onConfirm: onAccept
            )
        case .read:
            CleansiaPrimaryButton(L10n.close, action: onClose)
        }
    }
}

private struct JobFactsCard: View {
    let facts: WorkContractJobFacts
    let locale: Locale

    private var window: String {
        OrdersFormat.window(facts.cleaningDateTimeUtc, minutes: facts.estimatedMinutes, locale: locale)
    }

    private var orderNumber: String? {
        guard let number = facts.orderNumber, !number.isBlank else { return nil }
        return number
    }

    private var scopeLine: String? {
        let parts = [
            facts.rooms > 0 ? OrdersFormat.rooms(facts.rooms) : nil,
            facts.bathrooms > 0 ? OrdersFormat.baths(facts.bathrooms) : nil
        ].compactMap { $0 }
        return parts.isEmpty ? nil : parts.joined(separator: " · ")
    }

    private var contentsLine: String? {
        let names = facts.packages + facts.services + facts.extraSlugs.map(OrderExtras.name)
        return names.isEmpty ? nil : names.joined(separator: ", ")
    }

    var body: some View {
        VStack(alignment: .leading, spacing: Spacing.xxs) {
            HStack(alignment: .firstTextBaseline) {
                Text(orderNumber.map(L10n.WorkContract.orderNumber) ?? window)
                    .font(CleansiaTypography.titleMedium)
                    .foregroundColor(CleansiaColors.onSurface)
                    .frame(maxWidth: .infinity, alignment: .leading)
                Text(EarningsFormat.wholeMoney(facts.totalPrice, currencyCode: facts.currencyCode))
                    .font(CleansiaTypography.titleMedium)
                    .foregroundColor(CleansiaColors.primary)
            }
            if orderNumber != nil {
                FactLine(window)
            }
            if let location = facts.locationApproximate, !location.isBlank {
                FactLine(location)
            }
            if let scopeLine {
                FactLine(scopeLine)
            }
            if let contentsLine {
                FactLine(contentsLine)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(Spacing.s)
        .background(CleansiaColors.surfaceVariant.opacity(0.5), in: RoundedRectangle(cornerRadius: CornerRadius.small))
    }
}

private struct FactLine: View {
    let text: String

    init(_ text: String) {
        self.text = text
    }

    var body: some View {
        Text(text)
            .font(CleansiaTypography.bodyMedium)
            .foregroundColor(CleansiaColors.onSurfaceVariant)
    }
}

private struct AcceptanceFacts: View {
    let acceptance: WorkContractAcceptanceFacts
    let renderedLanguage: String?
    let locale: Locale

    /// Named only when the accepted text is not the one on screen, so the reader knows which string
    /// the acceptance binds.
    private var acceptedLanguageName: String? {
        guard let accepted = acceptance.acceptedLanguage, !accepted.isBlank,
              let rendered = renderedLanguage, accepted.caseInsensitiveCompare(rendered) != .orderedSame
        else { return nil }
        return locale.localizedString(forLanguageCode: accepted) ?? accepted
    }

    var body: some View {
        VStack(alignment: .leading, spacing: Spacing.xxs) {
            FactLine(L10n.WorkContract.acceptedOn(
                OrdersFormat.dateTime(acceptance.acceptedOn, locale: locale),
                acceptance.documentVersion
            ))
            if let acceptedLanguageName {
                FactLine(L10n.WorkContract.acceptedInLanguage(acceptedLanguageName))
            }
        }
    }
}

private struct NoticeRow: View {
    let text: String

    var body: some View {
        HStack(alignment: .top, spacing: Spacing.xs) {
            Image(systemName: "info.circle")
                .foregroundColor(CleansiaColors.warningStar)
            Text(text)
                .font(CleansiaTypography.labelMedium)
                .foregroundColor(CleansiaColors.onSurface)
                .fixedSize(horizontal: false, vertical: true)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(Spacing.s)
        .background(CleansiaColors.warningStar.opacity(0.12), in: RoundedRectangle(cornerRadius: CornerRadius.small))
    }
}

private struct SheetMessage: View {
    let text: String
    let ctaLabel: String
    let onCta: () -> Void

    var body: some View {
        VStack(spacing: Spacing.m) {
            Text(text)
                .font(CleansiaTypography.bodyLarge)
                .foregroundColor(CleansiaColors.onSurfaceVariant)
                .multilineTextAlignment(.center)
            CleansiaPrimaryButton(ctaLabel, size: .medium, action: onCta)
                .fixedSize()
        }
        .padding(Spacing.xl)
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }
}

#if DEBUG
    extension WorkContract {
        static let preview = WorkContract(
            legalDocumentTextId: "text-1",
            version: "2026-09-20",
            language: "en",
            title: "Contract for work",
            contentHtml: "<p>This contract for work is concluded between the customer and the cleaner.</p>",
            facts: WorkContractJobFacts(
                orderNumber: "CL-2026-0042",
                cleaningDateTimeUtc: Date(timeIntervalSince1970: 1_786_200_000),
                estimatedMinutes: 180,
                totalPrice: 1850,
                currencyCode: "CZK",
                locationApproximate: "Praha 4 · 14000",
                rooms: 3,
                bathrooms: 1,
                services: ["Standard cleaning"],
                packages: [],
                extraSlugs: ["inside-oven"]
            ),
            acceptance: nil
        )
    }

    struct WorkContractSheet_Previews: PreviewProvider {
        static var previews: some View {
            Group {
                WorkContractSheetContent(request: .take(orderId: "order-1"), state: .loaded(.preview))
                    .previewDisplayName("Take")
                WorkContractSheetContent(
                    request: .accept(orderId: "order-1"),
                    state: .loaded(.preview),
                    notice: .textUpdated
                )
                .previewDisplayName("Accept · text updated")
                WorkContractSheetContent(request: .read(acceptanceId: "acc-1"), state: .loading)
                    .previewDisplayName("Loading")
                WorkContractSheetContent(
                    request: .take(orderId: "order-1"),
                    state: .unavailable(ApiError(code: WorkContractErrorKey.documentNotFound))
                )
                .previewDisplayName("Unavailable")
            }
            .background(CleansiaColors.surface)
        }
    }
#endif
